### ISSUE TITLE
Azure::Core::Context::IsCancelled() performs 1/3 times slower than .NET CancellationToken.IsCancellationRequested 

### ISSUE THREAD

**Issue by vhvb1989** [Azure/azure-sdk-for-cpp#1589]
while trying to use `Azure::Core::Context` for the performance framework, I noticed C++ behaving at 1/3 speed compared to .NET.

While running the same test with no implementation on the same machine and with the same configuration, I got the next resutls:
.NET -> ~30Million operations/sec
C++ -> ~10Million operations/sec 

(Both compiled on release mode and fully optimized)

Then I start triaging the C++ application, searching for top performance offenders. As part of the experiments, I update the C++ implementation and test adding/removing the next components:
- static inline vs inline vs no online methods.  No significant changes.
- Run test without reporting progress. No significant changes.
- Run test without recording times (no calls to `std::chrono::system_clock::now()` and no time conversion). No significant changes.
- Run test without calling context.IsCancelled(). `Significant` changes. The test results are similar to .Net (~30 Million op/sec).

After looking into `context.IsCancelled()` implementation. A context makes a lineal validation to all the linked list (parents) and calls `CancelWhen()` on each context

```cpp
time_point Azure::Core::Context::CancelWhen() const
{
  auto result = time_point::max();
  for (auto ptr = m_contextSharedState; ptr; ptr = ptr->Parent)
  {
    auto cancelAt = ContextSharedState::FromMsecSinceEpoch(ptr->CancelAtMsecSinceEpoch);
    if (result > cancelAt)
    {
      result = cancelAt;
    }
  }

  return result;
}
```

@JeffreyRichter  The first question would it be. Do we want to have a context that is as fast as the .NET ?
If yes, we can think about some ideas to improve the cancellation strategy. For example, setting the expected timepoint for the context to be cancelled instead of calculating it on every `IsCancelled()` call.

Consider the next sample application
```cpp

// create a context with deadline (5 sec)
auto deadLineSeconds = std::chrono::seconds(5);
auto contextWithTimeout = Azure::Core::GetGlobalContext.WithDeadline(deadLineSeconds);

// run the task with that context
auto task = std::thread([&contextWithTimeout](){
  while(!contextWithTimeout.IsCancelled()) {
   // do nothing  
  }
}

task.join()

```

Now take a look to the next sample which performs 3 times faster than the previous
```cpp
// create a cancellation reference
bool isCancelled = false;
// Set 5 sec timeout
auto deadLineSeconds = std::chrono::seconds(5);

// run the task and use one extra thread to update cancellation
auto task = std::thread([&isCancelled, &deadLineSeconds  ](){
  
   // Create thread for cancelling
  auto cancelThread = std::thread([&isCancelled, &deadLineSeconds]() {
    // just sleep until the deadline
    std::this_thread::sleep_for(deadLineSeconds);
    isCancelled = true;
  }

  while(!isCancelled) {
   // do nothing  
  }

  cancelThread.join();
}

task.join()

```

The second example is currently the workaround for the performance framework.
The downside of this approach is: The entire program can't be cancelled from the main parent context (like the global context) when the while() method from previous sample is running.

**Comment by JeffreyRichter** [comment]
I'm not worried about the perf of Context because it is generally called once per retry and while getting a block of bytes from the stream to upload/download. So, it's only used in the context (pun intended) of network operations.

That being said, if we can improve its perf easily, we should. Maybe we can remove the call to ContextSharedState::FromMsecSinceEpoch inside the loop.

Another option is to have IsCancelled run the same loop but exit the loop early as soon as a timeout is found - this way the whole linked-list doesn't have to be walked.

Also, you don't say how many Context objects are in the linked-list for your perf test. I would expect most customers to have no more than 10 and even 5 is probably more common.

Also, our Context does way more than .NET's CancellationToken. I'm happy to pay a bit more perf price for our additional features.

I did file this bug issue related to Context 2 days ago: https://github.com/Azure/azure-sdk-for-cpp/issues/1537.
I really think this should be done it might improve for allocating/initializing Context objects and improve perf when looking for key/value pairs but not improve perf when checking for cancellation.

**Comment by vhvb1989** [comment]
> Also, you don't say how many Context objects are in the linked-list for your perf test.

Just one context. Created from the root context.

I looked closer and the issue is not because of the lineal check through the linked list. The issue is that the `context` uses an atomic variable in its implementation, which means there is a performance penalty we have to pay for the thread sync/lock anytime we check a context for cancellation.

I noticed that the atomic value `std::atomic_int64_t CancelAtMsecSinceEpoch;` is set on constructor and updated when calling `Cancel()`.  I wonder if we can have an atomic boolean for this instead and let the `CancelAtMsecSinceEpoch` state to be just a constant that is used to check for cancellation.
